snippets/ModalDialog.md

```diff
@@ -0,0 +1,142 @@
+
+After the component mounted,Dialog binds a custom event called `modal` to document element.The event handler defined in our component named `modalHandler`.It accepts and renders the data passed by users.When document node captured a `modal` event,`modalHandler` works.
+
+Before the component unmounted,Dialog removes the custom event from document element in case of memory leaks.
+
+Dialog exposed the `show` method via using `static` reserved keyword so that we can call it through Dialog class.The `show` method accepts a object parameter that contains three properties: `title`, `closeOnClick`, and `content`.
+
+* `title` is the text you wanna show as the title of the dialog,passed as a String.
+* `closeOnClick` is a boolean to determine if the modal should close when clicked on(true), or only when the X button is pressed(false).
+* `content` is a jsx element.Dialog will render it.You can pass any content you want to show in the dialog.
+
+See the live demo on [stackblitz.com](https://stackblitz.com/edit/react-7yg2gr).
+
+```css
+  .modal {
+    position: fixed;
+    top: 0;
+    left: 0;
+    width: 100%;
+    height: 100%;
+    background-color: rgba(0, 0, 0, 0.6);
+    z-index: 9998;
+    display: flex;
+    justify-content: center;
+    align-items: center;
+  }
+  .dialog {
+    background-color: white;
+    border-radius: 5px;
+    overflow: hidden;
+  }
+  .dialog-title {
+    box-sizing: border-box;
+    width: 100%;
+    height: 48px;
+    padding: 0 16px;
+    border-bottom: 0.5px solid #c3c3c3;
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
+  }
+  .dialog-close {
+    font-size: 32px;
+    color: #c3c3c3;
+    cursor: pointer;
+    transform: rotate(45deg);
+    user-select: none;
+  }
+  .dialog-close:hover {
+    color: red;
+  }
+  .dialog-content {
+    min-width: 300px;
+  }
+```
+
+```jsx
+class Dialog extends React.Component {
+  constructor() {
+    super();
+    this.modalHandler = (e) => {
+      this.setState({
+        data: e.detail.data,
+        visible: true
+      });
+    };
+    this.state = {
+      data: {
+        title: '',
+        closeOnClick: false,
+        content: ''
+      },
+      visible: false
+    };
+    this.close = this.close.bind(this);
+    this.modalClick = this.modalClick.bind(this);
+  }
+  render() {
+    return <div className="modal" onClick={this.modalClick} style={{ display: this.state.visible ? '' : 'none'}}>
+      <div className="dialog">
+        <div className="dialog-title">{ this.state.data.title }<span className="dialog-close" onClick={this.close}>+</span></div>
+        <div className="dialog-content">
+          {
+            this.state.data.content
+          }
+        </div>
+      </div>
+    </div>
+  }
+  componentDidMount() {
+    document.addEventListener('modal', this.modalHandler);
```

Comment:
Usage of refs instead of document object.

Reply:
A ref is used for interacting with the DOM through the React instance and should only be used for certain things. This is the exact opposite of what we want. This is not a use case it should be used for as you have no reason to react with the DOM, you only need to react with the current instance of the Modal. This is why we register an event listener. If you registered it on the node, you'd have to find that node EVERY TIME you wanted to show the modal, and if your default state is to return null if there is no render(a suggestion I'm considering making after some review of design best practices), you have NO ref and thus no way to fire. By putting it on the document as a global event listener we can easily fire for it and provide react tools to show and use it. In fact there are still some more changes that could be made to make this even slicker

Reply:
I think I wasn't able to explain this clearly before. So, as you already know, `document.addEventListener` is the usual way of attaching the event listeners or interacting with the dom when we build websites. But when it comes to react ( and mostly when you're working with server-side rendered applications ), this implementation would raise an error if you're trying to modify something. It is not a problem with implementation as it is quite logical and you would expect it to work ( and it works too - not sure about prod but in development it does ) but it will fail. Let's say you take this example and run this on your local development server. It'll work great. Now let's say this code example is used in a production website. Even the local tests would pass but if you're using some vendor service like circle ci or travis you would see that the tests fail. 
I could provide an example but right now I'm just packed with a bunch of stuff at office. My suggestion was to use `ref.current` ( using `createRef` and not callback ref - since callback ref is updated twice which we don't need ) instead of `document` object since it ensures the code doesn't fail you or anyone who uses this. 
Anyways not a big deal since anyone who'd run into this kind of error will obviously switch to the reactive way like I did when I had first encountered this.

Reply:
Note about tonality: There is a possibility this post could sound quite sarcastic, that is not my intent. I'm merely confused by what you're attempting to tell us.

If I'm reading this correctly you're suggesting we instead use a `React.creatRef()` call, and then expose it, breaking component encapsulation?

Or are you suggesting we use a `React.createRef()` and then encapsulate the rest of the page in the modal? Why not just instead expose a method the children would use to update the state?

Otherwise I'm all sorts of lost and have a lot of problems with your argument. Yes React server side rendering is a thing with `react-dom-server`, but those are meant for static markup. You can then use `react-dom.hydrate()` to mount, and the mount event will fire and give proper event handling as long as you've set up your life cycle methods correctly. At that point you would use `componentWillMount` instead of `componentDidMount`(maybe we should change to that or an invisible `<div />`?).

As for your argument with testing, what sort of testing are you doing that you're not properly mocking or implementing the DOM APIs in your test file if you're going to use them? And then your integration tests should most definitely be against a headless browser with an interaction client firing the events you're expecting the DOM interaction to fire if you want to go that far.

Otherwise I'm quite lost by your example of changes. I say this with three and a half years as a React developer who first looked at the library for work purposes on it's initial release as a replacement for Polymer and jQuery.

Maybe there is something I'm missing, but without clarification or a better example I'm not picking up what you're putting down about this client side rendering library.

Reply:
I guess you're still not quite clear with what I said. 
And on a side note, no I am not suggesting that we expose the component ref ( which is a nasty idea ). Parent can control the child through the exposed method. I was suggesting that we use `createRef` inside the modal component to attach and remove event listeners or any kind of interaction / manipulation to the DOM. 
And on the testing part, when I bumped into this problem, I guess tests were written in enzyme and mocha ( for unit testing ) although I don't quite remember it. And by far with integration testing ( selenium ), I've never run into problems ( maybe because I don't use vanilla to interact / manipulate the DOM in my react components ).
And again, no problem with this code being kept as it is. Whosoever runs into the problem would try the suggestion that I gave ( since no other way ). 

And also, there was no problem with the tone. It's good to have healthy discussions on weird problems everyone faces at their workplace on some stage in the dev cycle. 😄

Reply:
Ah I think I get it now. In tests I would have just mocked the `document.addEventListener()` to add it to a map to be called for testing purposes in my scripts(which I've done for this exact case before)

I would love to see an example of how you would do the modal firing though since I'm not exactly certain how you could keep encapsulation with your suggestion...

Reply:
Well if i had to, I would let the parent be in complete control, something like this : 
```jsx
const Dialog = (props) => {
    return (
    !props.open ? null : <div className="modal">
        <div className="dialog">
            <div className="dialog-title">
                 {props.title}
                 <span className="dialog-close" onClick={props.onClose}>+</span>
            </div>
            <div className="dialog-content">
                {
                    props.content
                }
            </div>
        </div>
    </div>
    )
}
```
This removes the need for adding/removing event listeners or using refs altogether. Simple and straightforward. Parent keeps the track of modal state and also has complete control of whatever needs to be done before / after modal opens / closes.

Reply:
Well if I had to, I would let the parent be in complete control, something like this : 
```jsx
const Dialog = (props) => {
    return (
    !props.open ? null : <div className="modal">
        <div className="dialog">
            <div className="dialog-title">
                 {props.title}
                 <span className="dialog-close" onClick={props.onClose}>+</span>
            </div>
            <div className="dialog-content">
                {
                    props.content
                }
            </div>
        </div>
    </div>
    )
}
```
This removes the need for adding/removing event listeners or using refs altogether. Simple and straightforward. Parent keeps the track of modal state and also has complete control of whatever needs to be done before / after modal opens / closes.

Reply:
As of React 16 [portals](https://reactjs.org/docs/portals.html) were introduced for exactly this use case among others apparently.

I'd say we either want to move to that, or keep the event version(testing it in jest is very straightforward and easy, [just mock the APIs you'll be calling with jest](https://medium.com/@DavideRama/testing-global-event-listener-within-a-react-component-b9d661e59953))

@rishichawda I have an issue with the version you've suggested in that it destroys the DOM tree with styling, which could cause a HUGE render gap either by styling shoving things in weird places, or unforeseen performance due to crawl. It also requires the parent to cascade the function down and means there is developer debt in remembering to pass a function from parent to child to child to child to child...

By nesting at a "root" level in the app(near the top, as close to the body root as possible) we can avoid styling shoving elements around and don't have to worry about style collisions due to inheritance other than on the body. It's not that it's a bad thing, it's that it is a dangerous thing to nest your modals. It's why the 'portals' idea was adopted by MANY people, and our event model basically is our portal without requiring React 16(I've used this model since 14).

So I'd say at this point there are three options:

1. Event based modal
2. React child based modal
3. Portal based modal

I'm leaning towards 1 or 3 since 2 requires extra logic beyond just "display modal"(even if it's possibly as small as 'pass firing function') and more developer debt beyond 'import, call static' or 'render portal'.

I like the event based model though because of an above point: It doesn't require the latest(and is compatible with many versions) and doesn't incur massive developer debt. We might just want to help people with their testing since this seems to be a point of contest.

I'll be posting a copy of this comment in the main PR

Reply:
> @rishichawda I have an issue with the version you've suggested in that it destroys the DOM tree with styling, which could cause a HUGE render gap either by styling shoving things in weird places, or unforeseen performance due to crawl. 

Hmm didn't think about that. Nice @skatcat31 . :smiley: